app/app/src/main/kotlin/com/hedvig/android/app/ui/DeepLinkFirstUriHandler.kt

```diff
@@ -0,0 +1,26 @@
+package com.hedvig.android.app.ui
+
+import androidx.compose.ui.platform.UriHandler
+import androidx.core.net.toUri
+import androidx.navigation.NavController
+import com.hedvig.android.logger.logcat
+
+/**
+ * Tries to open a deep link with the [navController] and falls back to the [delegate] if the uri is not a valid deep
+ * link to our app
+ */
+internal class DeepLinkFirstUriHandler(
+  private val navController: NavController,
+  private val delegate: UriHandler,
+) : UriHandler {
+  override fun openUri(uri: String) {
+    if (navController.graph.hasDeepLink(uri.toUri())) {
+      logcat { "DeepLinkFirstUriHandler will try to navigate to uri:$uri" }
+      navController.navigate(uri.toUri())
+      logcat { "DeepLinkFirstUriHandler did deep link to uri:$uri" }
+    } else {
+      logcat { "DeepLinkFirstUriHandler falling back to default UriHandler for uri:$uri" }
+      delegate.openUri(uri)
```

Comment:
So if I understood correctly, this delegate (which is LocalUriHandler.current in our case) when opening a link, creates a separate backstack, and this causes all the problems, right? So you avoid this by calling navController directly first, which should in reality cover all cases with our deep links?

Reply:
Yeap that is exactly the problem.
It kinda opens our app as if we were coming from a deep link from another app. I am not sure how this works exactly, but then when we end up in the deep link, we have no backstack behind us, we are just in our app in a new "Task" I think it's called, and we lose track of what was there before.
And yes exactly, if it's our own deep link this should just work instantly and it's all good. If not it will in any case go and try to open another app or a website. Exactly what we want here!